File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/ReactiveStore.java
package uk.co.victoriajanedavis.chatapp.domain;


import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;

/**
 Interface for any kind of reactive store.
 */
public interface ReactiveStore<Key, Value> {

    Completable storeSingular(@NonNull final Value model);

    Completable storeAll(@NonNull final List<Value> modelList);

    Completable replaceAll(@Nullable final Key key, @NonNull final List<Value> modelList);

    Completable delete(@NonNull final Value model);

    Observable<Value> getSingular(@Nullable final Key key);

    Observable<List<Value>> getAll(@Nullable final Key key);
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/ReactiveStore.kt
package uk.co.victoriajanedavis.chatapp.domain


import io.reactivex.Completable
import io.reactivex.Observable

/**
 * Interface for any kind of reactive store.
 */
interface ReactiveStore<Key, Value> {

    fun storeSingular(value: Value): Completable

    fun storeAll(valuesList: List<Value>): Completable

    fun replaceAll(key: Key?, valuesList: List<Value>): Completable

    fun delete(value: Value): Completable

    fun getSingular(key: Key?): Observable<Value>

    fun getAll(key: Key?): Observable<List<Value>>
}
